File: talks/matlab_vs_python/bloch/bloch.m
% Plot the pulse sequence
% create figure
figure(); 

% get pulse sequence B-fields from 0-5 ms
pulseq = zeros(1000,3);
for i = 1:1000
    pulseq(i,:) = B_eff(i*0.005); 
end

% plot RF
subplot(2,1,1);
plot(pulseq(:,1));
ylabel('B1');

% plot gradient
subplot(2,1,2);
plot(pulseq(:,3));
ylabel('Gradient');

%% Integrate ODE
T1 = 1500;
T2 = 50;
t0 = 0;
t1 = 5;
dt = 0.005;
M0 = [0; 0; 1];
[t, M] = ode45(@(t,M)bloch_ode(t, M, T1, T2), linspace(t0, t1, (t1-t0)/dt), M0);

%% Plot Results
% create figure
figure();hold on;

% plot x, y and z components of Magnetisation
plot(t, M(:,1), 'linewidth', 2);
plot(t, M(:,2), 'linewidth', 2);
plot(t, M(:,3), 'linewidth', 2);

% add legend and grid
legend({'Mx','My','Mz'});
grid on;

%% define the bloch equation
function dM = bloch_ode(t, M, T1, T2)
    % get effective B-field at time t
    B   =   B_eff(t);                               

    % cross product of M and B, add T1 and T2 relaxation terms
    dM  =  [M(2)*B(3) - M(3)*B(2) - M(1)/T2;        
            M(3)*B(1) - M(1)*B(3) - M(2)/T2;        
            M(1)*B(2) - M(2)*B(1) - (M(3)-1)/T1];   
end

%% define effective B-field
function b = B_eff(t)
    % Do nothing for 0.25 ms
    if t < 0.25             
        b = [0, 0, 0];
    % Sinc RF along x-axis and slice-select gradient on for 1.00 ms
    elseif t < 1.25             
        b = [pi*sinc((t-0.75)*4), 0, pi];
    % Do nothing for 0.25 ms
    elseif t < 1.50             
        b = [0, 0, 0];
    % Slice refocusing gradient on for 1.50 ms
    % Half the area of the slice-select gradient lobe
    elseif t < 3.00             
        b = [0, 0, -(1/3)*pi];
    % pulse sequence finished
    else                       
        b = [0, 0, 0];
    end
end

File: talks/matlab_vs_python/bloch/bloch.py
#%% Imports
import numpy as np
from scipy.integrate import ode
import matplotlib.pyplot as plt

#%% Define bloch and B_eff functions
def bloch_ode(t,M,T1,T2):
    B = B_eff(t)
    return np.array([M[1]*B[2] - M[2]*B[1] - M[0]/T2,
                     M[2]*B[0] - M[0]*B[2] - M[1]/T2,
                     M[0]*B[1] - M[1]*B[0] - (M[2]-1)/T1])

def B_eff(t):
    if t < 0.25:
        return np.array([0, 0, 0])
    elif t < 1.25:
        return np.array([1.8*np.sinc(t-0.75), 0, 0])
    elif t < 1.50:
        return np.array([0, 0, 0])
    elif t < 3.00:
        return np.array([0, 0, 2*np.pi])
    else:
        return np.array([0, 0, 0])

#%% Integrate ODE
t = np.array([0])
M = np.array([[0, 0, 1]])
dt= 0.005
r = ode(bloch_ode)\
   .set_integrator('dopri5')\
   .set_initial_value(M[0],t[0])\
   .set_f_params(1500, 50)
while r.successful() and r.t < 5:
    t = np.append(t,r.t+dt)
    M = np.append(M, np.array([r.integrate(t[-1])]),axis=0)

#%% Plot Results
plt.plot(t,M[:,0])
plt.plot(t,M[:,1])
plt.plot(t,M[:,2])